src/net/socket_state.rs

```diff
@@ -94,28 +100,40 @@ impl SocketState {
         })
     }
 
-    // Regularly checks the last_heard attribute of all the connections in the manager to see if any have timed out
-    fn check_for_timeouts(&mut self) {
+    // This function starts a background thread that does the following:
+    // 1. Gets a read lock on the HashMap containing all the connections
+    // 2. Iterate through each one
+    // 3. Check if the last time we have heard from them (received a packet from them) is greater than the amount of time considered to be a timeout
+    // 4. If they have timed out, send a notification up the stack
+    fn check_for_timeouts(&mut self) -> Result<thread::JoinHandle<()>> {
         let connections_lock = self.connections.clone();
         let sleepy_time = Duration::from_secs(self.timeout);
         let poll_interval = Duration::from_secs(TIMEOUT_POLL_INTERVAL);
 
-        thread::Builder::new()
+        Ok(thread::Builder::new()
             .name("check_for_timeouts".into())
             .spawn(move || loop {
-                let connections = connections_lock.read().expect("Unable to aquire read lock");
-
-                for (key, value) in connections.iter() {
-                    let connection = value.read().expect("Unable to aquire read lock");
-                    let last_heard = connection.last_heard();
-                    if last_heard >= sleepy_time {
-                        // TODO: pass up client TimedOut event
-                        error!("Client has timed out: {:?}", key);
+                {
```

Comment:
Are you sure you need this extra scope here? The loop scope should cover the dropping of the lock.